kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala

```diff
@@ -0,0 +1,238 @@
+
+    def fold[B](andCase: (B, B) => B, toCase: (B, B) => B, valueCase: A => B, emptyCase: => B): B =
+        this match
+            case And(left, right) =>
+                andCase(left.fold(andCase, toCase, valueCase, emptyCase), right.fold(andCase, toCase, valueCase, emptyCase))
+            case To(left, right) =>
+                toCase(left.fold(andCase, toCase, valueCase, emptyCase), right.fold(andCase, toCase, valueCase, emptyCase))
+            case Value(value) => valueCase(value)
+            case Empty        => emptyCase
+end LayerLike
+
+object LayerLike:
+    def apply[A](value: A): LayerLike[A]            = Value(value)
+    given [A]: CanEqual[LayerLike[A], LayerLike[A]] = CanEqual.derived
+
+final case class Node[Key, Value](inputs: Set[Key], outputs: Set[Key], value: Value)
+object Node:
+    given [Key, Value]: CanEqual[Node[Key, Value], Node[Key, Value]] = CanEqual.derived
+
+enum Validated[+E, +A]:
+    case Success(value: A)
+    case Error(errors: ::[E])
+
+    def map[B](f: A => B): Validated[E, B] =
+        this match
+            case Success(value) => Success(f(value))
+            case Error(errors)  => Error(errors)
+
+    def flatMap[E1 >: E, B](f: A => Validated[E1, B]): Validated[E1, B] =
+        this match
+            case Success(value) => f(value)
+            case Error(errors)  => Error(errors)
+
+    def zipWith[E1 >: E, B, C](that: Validated[E1, B])(f: (A, B) => C): Validated[E1, C] =
+        (this, that) match
+            case (Success(value1), Success(value2)) => Success(f(value1, value2))
+            case (Error(e :: es), Error(es2))       => Error(::(e, es ++ es2))
+            case (Error(errors1), _)                => Error(errors1)
+            case (_, Error(errors2))                => Error(errors2)
+end Validated
+
+object Validated:
+    def succeed[E, A](value: A): Validated[E, A] = Success(value)
+    def error[E, A](error: E): Validated[E, A]   = Error(::(error, Nil))
+
+    def traverse[E, A, B](set: Set[A])(f: A => Validated[E, B]): Validated[E, Set[B]] =
+        set.foldLeft[Validated[E, Set[B]]](Validated.Success(Set.empty)) { (acc, item) =>
+            acc.zipWith(f(item)) { (accSet, b) =>
+                accSet + b
+            }
+        }
+
+    def sequence[E, A](set: Set[Validated[E, A]]): Validated[E, Set[A]] =
+        set.foldLeft(Validated.Success(Set.empty)) { (acc, item) =>
+            acc.zipWith(item) { (accSet, b) =>
+                accSet + b
+            }
+        }
+end Validated
+
+enum GraphError[Key, Value]:
+    case MissingInput(input: Key, parent: Option[Node[Key, Value]])
+    case CircularDependency(node: Node[Key, Value]) // TODO: Maybe also trace the path.
+    case AmbiguousInputs(target: Key, found: Set[Node[Key, Value]], parent: Option[Node[Key, Value]])
+end GraphError
+
+final case class Graph[Key, Value](nodes: Set[Node[Key, Value]])(equals: (Key, Key) => Boolean):
```

Comment:
@kitlangton should we make the whole graph/validated private? Is there any reason we can't?

Reply:
Agree! We could move impls to a companion object to avoid polluting the `kyo.internal` namespace as well.